File: Practica_45.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Scanner;

public class Practica_45 {

    /**
     * Este metodo pide por teclado los datos necesarios para agrega un nuevo usuario al txt y lo agrega
     * @param scanner El scannner que se inicia en la funcion menu()
     * @param fichero La ruta del fichero
     */
    public static void agregar(Scanner scanner , String fichero){
        System.out.println("Introduzca el dni: ");
        String dni = scanner.nextLine();
        System.out.println("Introduzca el nombre: ");
        String nombre = scanner.nextLine();
        System.out.println("Introduzca los apellidos: ");
        String apellidos = scanner.nextLine();
        System.out.println("Introduzca el correo: ");
        String correo = scanner.nextLine();
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(fichero, true));){
            String linea = (dni + ", " + nombre + ", " + apellidos + ", " + correo);
            writer.write(linea);
            writer.newLine();
        }
        catch (IOException e) {
            System.out.print("Error al leer el archivo");
        }
    }
    
    /**
     * Este metodo borra un usuario mediante el dni pedido por teclado
     * @param scanner El scannner que se inicia en la funcion menu()
     * @param fichero La ruta del fichero
     */
    public static void eliminar(Scanner scanner , String fichero){
        // se crea un archivo temporal
        File arhchivoOriginal = new File(fichero);
        File archivoTemp = new File("temp.txt");
        System.out.println("Introduzca el dni: ");
        String dni = scanner.nextLine();
        String linea = null;
        try (FileReader fileReader = new FileReader(fichero); BufferedReader bufferedReader = new BufferedReader(fileReader); 
        BufferedWriter writer = new BufferedWriter(new FileWriter("temp.txt"))){
        // se escribe en el archivo temporal todo el contenido del original menos la linea de la que coincida el dni
        while((linea=bufferedReader.readLine()) != null) {
            if (!linea.substring(0, 9).equals(dni)){
                writer.write(linea);
                writer.newLine();
            }
        }  
        }
        catch (IOException e) {
            System.out.print("Error al leer el archivo");
        }
        // se borra el archivo original y se renombra el archivo temporal
        arhchivoOriginal.delete();
        archivoTemp.renameTo(arhchivoOriginal);
    }
    

    /**
     * Este metodo pide el dni por teclado , busca en el fichero la linea con ese dni , y devuelve la linea
     * @param scanner El scannner que se inicia en la funcion menu()
     * @param fichero La ruta del fichero
     * @return 
     */
    public static String busqueda_dni(Scanner scanner , String fichero){
        System.out.println("Introduzca el dni: ");
        String dni = scanner.nextLine();
        String linea = null;
        try (FileReader fileReader = new FileReader(fichero);
        BufferedReader bufferedReader = new BufferedReader(fileReader);){
        while((linea=bufferedReader.readLine()) != null) {
            if (linea.substring(0, 9).equals(dni)){
                return linea;
            }
        }
        System.out.print("El dni introducido no existe");
        return linea;
        }
        catch (IOException e) {
            System.out.print("Error al leer el archivo");
            return linea;
        }
    }

    /**
     * Este metodo imprime los datos de la linea que se le pasa
     * @param linea La linea que va a imprimir
     */
    public static void imprimir_datos(String linea){
        String[] linea_c = linea.split(", ");
        System.out.println("nombre: " + linea_c[1]);
        System.out.println("apellidos: " + linea_c[2]);
        System.out.println("correo: " + linea_c[3]);
    } 


    /**
     * El metodo menu es el metodo que se ejecuta al ejecutar el programa y es un menu que le permite al usuario usar las funcionalidades del programa
     */
    public static void menu(){
        Scanner scanner = new Scanner(System.in);
        while (true){
            String fichero = "usuarios.txt";
            System.out.println("\n--- Opciones: ---");
            System.out.println("1. Buscar usuario por DNI");
            System.out.println("2. Agregar usuario");
            System.out.println("3. Eliminar usuario");
            System.out.println("4. cerrar programa");
            System.out.println("Seleccione una opcion: ");
            String opcion = scanner.nextLine();
            if (opcion.equals("1")) {
                String linea = busqueda_dni(scanner , fichero);
                if (linea == null){
                    System.out.println("\nVuelva a elejir una opcion");
                }
                else {
                    imprimir_datos(linea);
                }
            }
            else if(opcion.equals("2")){
                agregar(scanner , fichero);
            }
            else if (opcion.equals("3")){
                eliminar(scanner , fichero);
            }
            else if(opcion.equals("4")){
                scanner.close();
                return;
            }
        }
    }

    public static void main(String[] args){
        menu();
    }
}

File: Practica 45.py
import os


def agregar():
    """
    Este metodo pide por teclado los datos necesarios para agrega un nuevo usuario al txt y lo agrega
    """
    with open('usuarios.txt' , 'a' , encoding='utf-8') as fichero:
        dni = str(input("Introduzca el dni: "))
        nombre = str(input("Introduzca el nombre: "))
        apellidos = str(input("Introduzca los apellidos "))
        correo = str(input("Introduzca el correo: "))
        linea = (dni + ", " + nombre + ", " + apellidos + ", " + correo)
        fichero.write(linea)

def eliminar(fichero):
    """
    Este metodo borra un usuario mediante el dni pedido por teclado
    """
    with open('temp.txt' , 'a' , encoding='utf-8') as fichero_escritura:
        dni = str(input("Introduzca el dni que quieres buscar: "))
        for linea in fichero:
            if dni != linea[:9]:
                fichero_escritura.write(linea)
        fichero.close()
        fichero_escritura.close()
        # se elemina el archivo original y se renombra el temporal
        os.remove('usuarios.txt')
        os.rename('temp.txt' , 'usuarios.txt')
        

def busqueda_dni(fichero):
    """
    Este metodo pide el dni por teclado , busca en el fichero la linea con ese dni , y devuelve la linea 
    """
    dni = str(input("Introduzca el dni que quieres buscar: "))
    for linea in fichero:
        # se establece un condicional que compara el dni introducido con el dni de cada linea del txt
        if dni == linea[:9]:
            return linea
    print("\nEl dni introducido no existe")

def imprimir_datos(linea):
    """
    Este metodo imprime los datos de la linea que se le pasa
    """
    linea_c = linea.split(",")
    nombre = linea_c[1]
    apellidos = linea_c[2]
    correo = linea_c[3]
    print("nombre: " , nombre)
    print("apellidos: " , apellidos)
    print("correo: " , correo)
    

def menu():
    """
    El metodo menu es el metodo que se ejecuta al ejecutar el programa y es un menu que le permite al usuario usar las funcionalidades del programa
    """
    while True:
        with open('usuarios.txt' , 'r' , encoding='utf-8') as fichero:
            print("\n--- Opciones: ---")
            print("1. Buscar usuario por DNI")
            print("2. Agregar registro")
            print("3. Eliminar registro")
            print("4. cerrar programa")
            opcion = input("Seleccione una opción: ")
            # si se elije la opcion 1 se llama al metodo de buscar dni y luego al metodo de imprimir los datos
            if opcion == '1':
                linea_actual = busqueda_dni(fichero)
                # se comprueba que la salida del metodo busqued_dni no sea null por haber introducido un dni inexistente
                if linea_actual != None:
                    imprimir_datos(linea_actual)
                else:
                    print("\nVuelva a elejir una opcion")
            elif opcion == '2':
                agregar()
            elif opcion == '3':
                eliminar(fichero)
            elif opcion == '4':
                fichero.close()
                return


menu()